ml-agents/mlagents/trainers/trainer/rl_trainer.py

```diff
@@ -131,9 +131,12 @@ def _maybe_write_summary(self, step_after_process: int) -> None:
         if step_after_process >= self.next_summary_step and self.get_step != 0:
             self._write_summary(self.next_summary_step)
 
-    def advance(self) -> None:
+    def advance(self, empty_queue: bool = False) -> None:
         """
         Steps the trainer, taking in trajectories and updates if ready.
+        Will block and wait briefly if there are no trajectories.
+        :param empty_queue: Whether or not to empty the queue when called. For synchronous
```

Comment:
Why not just always empty the queue when called?

Reply:
It's possible (e.g. in the case of very fast environments, or with enough parallel envs) that the queue fills faster than we can load it into the replay buffer. In that scenario the policy will never update since it'll be stuck in this loop. The opposite won't happen since the put() on the env_manager side is blocking, so if the queue is full the envs will stop while the trainers process the trajectories. 

We could possibly put the loop on the outside (i.e. include the policy update in the get loop). I'd still need some sort of option to disable the blocking behavior in the unthreaded case, though.

Reply:
@ervteng Makes sense. I did notice there was a limit on how many times you'll call `get` so I thought that issue might be taken care of in that way.

Reply:
Actually I think you're right - it would be. But it would severely rate-limit the number of updates in that case, e.g. if the queue maxlen is 1000, and its always full, we'd update at most with 1/1000 trajectories. Perhaps this issue isn't a huge deal, especially if the max queue length is small.

Reply:
I was just keeping an eye on potentially minimizing changes to our existing interfaces if possible.  Up to you on whether you want to worry about that.  The implementation as it is looks good for me.